### DISCUSSION TITLE
Why SD 5.5 use Thumbnail Decoding for large image to replace Downsizing in SD 4.0

### DISCUSSION THREAD

**Discussion by KittenYang** [SDWebImage/SDWebImage#3192] [General]
I noticed that SD 5.5 moved the Downsizing codes to `[SDImageCoderHelper decodedAndScaledDownImageWithImage:image limitBytes:bytes]; ` and not used.
Instead，SD 5.5.0 use the Thumbnail Decoding https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#thumbnail-decoding-550

So I write a test, both decode a large image, the limited size is 60M which means the  output image maximum size is small than (3960, 3960). Here is the result.

SD 4.x downsizing:
![image](https://user-images.githubusercontent.com/5517281/111853138-a6956980-8954-11eb-8f4a-aa2e8549eeb2.png)

SD 5.5 Thumbnail Decoding
![image](https://user-images.githubusercontent.com/5517281/111853145-ac8b4a80-8954-11eb-8b52-95548c1333cf.png)

Notice the peak memory, the Thumbnail Decoding workaround is way more higher than Downsizing workaround.

So why use Thumbnail Decoding instead of Downsizing.Do I miss some context?

**Comment by dreampiggy** [comment]
@KittenYang Hi. can you provide your image url, or any demo ?

The image bitmap buffer memory size, should be only related to image resolution.

For 3960 x 3960, it **MUST BE** `3960*3960*4` bytes, which is `59.820556640625 MB`.

If not, it's a bug, but not designed behavior. (The thumbnail decoding does not take effect at all ?)

**Reply by KittenYang** [reply]
@dreampiggy The Thumbnail Decoding does take effect. But the **Peak Memory** is still High. Look at the picture 2, the peak memory is 279.15M but picture 1 is only 52.69M. I uploaded a demo project you can test.

https://github.com/KittenYang/SDThumbnailDecodingPeakMemoryTest

Notice: please use *Instruments - Allocations* to monitor memory.

**Reply by KittenYang** [reply]
@dreampiggy Hi, have u tested it?
